services/venice-controller/src/main/java/com/linkedin/venice/controller/ProtocolVersionAutoDetectionService.java

```diff
@@ -0,0 +1,196 @@
+      while (!stop.get()) {
+        for (String clusterName: allClusters) {
+          if (!veniceParentHelixAdmin.isLeaderControllerFor(clusterName)) {
+            continue;
+          }
+          try {
+            // start the clock
+            long startTime = System.currentTimeMillis();
+            Long currentGoodVersion = getSmallestLocalAdminOperationProtocolVersionForAllConsumers(clusterName);
+            Long upstreamVersion = AdminTopicMetadataAccessor.getAdminOperationProtocolVersion(
+                veniceParentHelixAdmin.getAdminTopicMetadata(clusterName, Optional.empty()));
+            LOGGER.info(
+                "Current good Admin Operation version for cluster {} is {} and upstream version is {}",
+                clusterName,
+                currentGoodVersion,
+                upstreamVersion);
+            if (upstreamVersion != -1 && !Objects.equals(currentGoodVersion, upstreamVersion)) {
+              veniceParentHelixAdmin.updateAdminOperationProtocolVersion(clusterName, currentGoodVersion);
+              LOGGER.info(
+                  "Updated admin operation protocol version in ZK for cluster {} from {} to {}",
+                  clusterName,
+                  upstreamVersion,
+                  currentGoodVersion);
+            }
+            long elapsedTimeFromMsToMs = LatencyUtils.getElapsedTimeFromMsToMs(startTime);
+            getClusterStat(clusterName).recordProtocolVersionAutoDetectionLatencySensor(elapsedTimeFromMsToMs);
+          } catch (Exception e) {
+            LOGGER.warn("Received exception while running ProtocolVersionDetectionTask", e);
+            getClusterStat(clusterName).recordProtocolVersionAutoDetectionErrorSensor();
+          } catch (Throwable throwable) {
+            LOGGER.warn("Received a throwable while running ProtocolVersionDetectionTask", throwable);
+            getClusterStat(clusterName).recordProtocolVersionAutoDetectionThrowableSensor();
+          }
+        }
+      }
+    }
+
+    /**
+     * Get the smallest local admin operation protocol version for all consumers in the given cluster.
+     * This will help to ensure that all consumers are on the same page regarding the protocol version.
+     *
+     * @param clusterName The name of the cluster to check.
+     * @return The smallest local admin operation protocol version for all consumers (parent + child controllers)
+     * in the cluster.
+     */
+    public long getSmallestLocalAdminOperationProtocolVersionForAllConsumers(String clusterName) {
+      // Map to store all consumers versions
+      Map<String, Map<String, Long>> regionToControllerToVersionMap = new ConcurrentHashMap<>();
+
+      // Get all versions for parent clusters
+      Map<String, Long> parentClusterVersions =
+          veniceParentHelixAdmin.getVeniceHelixAdmin().getAdminOperationVersionFromControllers(clusterName);
+      regionToControllerToVersionMap.put("parentRegion", parentClusterVersions);
```

Comment:
Please move "parentRegion" to a static final string constant.

Reply:
Yes, move!